e3db/src/main/java/com/tozny/e3db/Client.java

```diff
@@ -1093,6 +1116,83 @@ public void run() {
     });
   }
 
+  /**
+   * Write the given file to E3DB. Intended for files from 5MB up to 5GB in size. The contents of the file are
+   * encrypted before being uploaded.
+   *
+   * @param type Type of the record. Cannot be {@code null} or blank.
+   * @param file Path to the file to upload. Cannot be {@code null}.
+   * @param plain Plaintext meta associated with the file. Can be {@code null}.
+   * @param handleResult Handles the result of the write.
+   */
+  public void writeFile(final String type, final File file, final Map<String, String> plain, final ResultHandler<RecordMeta> handleResult) {
+    checkNotEmpty(type, "type");
+    checkNotNull(file, "file");
+    if(plain != null && plain.size() > 0)
+      checkMap(plain, "plain");
+
+    onBackground(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          final byte[] ownAK = getOwnAccessKey(type);
+          Map<String, Object> meta = makeRecordMetaMap(type, plain);
+          File encryptedFile = Platform.crypto.encryptFile(file, ownAK);
+          Map<String, Object> fileMeta = makeFileMeta(encryptedFile);
+
+          Map<String, Object> record = new HashMap<>();
+          record.put("meta", meta);
+          record.put("file", fileMeta);
+
+          EXIT: {
+            retrofit2.Response<ResponseBody> postResponse = storageClient.writeFile(RequestBody.create(APPLICATION_JSON, mapper.writeValueAsString(record))).execute();
+            if (postResponse.code() != 202) {
+              uiError(handleResult, E3DBException.find(postResponse.code(), postResponse.message()));
+              break EXIT;
+            }
+
+            final JsonNode pendingFile = mapper.readTree(postResponse.body().string());
+            UUID pendingFileID = UUID.fromString(pendingFile.get("id").asText());
+            String destURL = pendingFile.get("signed_url").asText();
+
+            try {
+              Request postFile = new Request.Builder().url(destURL).post(RequestBody.create(APPLICATION_OCTET, encryptedFile)).build();
+              Response postFileResponse = anonymousClient.newCall(postFile).execute();
+              if (postFileResponse.code() != 200) {
+                uiError(handleResult, E3DBException.find(postFileResponse.code(), postFileResponse.message()));
```

Comment:
This may include some XML and other S3 nasties (from the response body), so not sure that should propagate up to the UI layer. https://github.com/tozny/e3db-java/blob/8dd8e3ffcee976635aaf96b1aadb7916f05fbb6b/e3db/src/main/java/com/tozny/e3db/E3DBException.java#L58 Will likely show the user the entire error which we may want to avoid.

Reply:
`postFileResponse.message()` is just the status line - so might be "Bad Request" or similar. The response body is not included.